Allow configuration of more SAML settings

alexmorrisnz opened the pull request:
Adds the ability to configure:
- Entity ID
- 'WantAssertionsSigned'
- ACS Bindings
- 'AllowCreate'

Allows you to match users on the nameid value

This allows us to use RealMe with auth_saml2

kabalin commented on classes/auth.php:
```diff
@@ -594,14 +596,19 @@ public function saml_login_complete($attributes) {
-                // We found a user.
-                break;
+        if ($this->config->matchnameid) {
+            $uid = $nameid->value;
```
Would be good to have test covering this case (the rest of `saml_login_complete` is test covered already).

alexmorrisnz replied:
Removed this option as nameidasattribute covers it

brendanheywood commented on lang/en/auth_saml2.php:
```diff
@@ -108,6 +108,16 @@
+$string['allowcreate'] = 'Allow Create';
+$string['allowcreate_help'] = '';
+$string['matchnameid'] = 'Match on NameID value';
+$string['matchnameid_help'] = '';
```
I think this definitely needs help text to distinguish it from nameidasattribute and explain when you would use one over the other, and possibly move them to be adjacent

alexmorrisnz replied:
I wasn't aware of nameidasattribute, seems to work fine with realme so will remove this setting

brendanheywood commented on lang/en/auth_saml2.php:
```diff
@@ -108,6 +108,16 @@
+$string['entityid'] = 'Entity ID';
+$string['entityid_help'] = '';
+$string['allowcreate'] = 'Allow Create';
+$string['allowcreate_help'] = '';
```
This definitely needs docs to tell it apart from 'autocreate' and I think both the autocreate and these docs should be refined something like:

allowcreate Allow creation of IdP users on demand
autocreate Allow creation of Moodle users on demand

brendanheywood commented on settings.php:
```diff
@@ -165,6 +165,54 @@
+        'auth_saml2/assertionsconsumerservices',
+        get_string('assertionsconsumerservices', 'auth_saml2'),
+        get_string('assertionsconsumerservices_help', 'auth_saml2'),
+        [
```
worth a micro refactor of array_keys($assertionsconsumerservices) ?

alexmorrisnz replied:
yep, done

brendanheywood commented on lang/en/auth_saml2.php:
```diff
@@ -108,6 +108,16 @@
+$string['assertionsconsumerservices'] = 'Assertions Consumer Services';
+$string['assertionsconsumerservices_help'] = '';
+$string['entityid'] = 'Entity ID';
+$string['entityid_help'] = '';
```
gonna sound like a broken record - this needs docs, none of these should be empty. This one I think its quite important to tell people that in 99.9% of cases they can leave it blank

alexmorrisnz replied:
yep added the help texts, should this be documented anywhere else?

danmarsden commented on lang/en/auth_saml2.php:
```diff
@@ -108,6 +108,14 @@
 $string['manageidpsheading'] = 'Manage available IdPs';
 $string['mdlattr_help'] = 'Which Moodle user field should the IdP attribute be matched to?';
 $string['mdlattr'] = 'Mapping Moodle';
+$string['wantassertionssigned'] = 'Want Assertions Signed';
```
Moodle uses sentence case - not camel case - also check the other strings like consumer services...

alexmorrisnz replied:
done

danmarsden commented on settings.php:
```diff
@@ -165,6 +165,42 @@
+        'auth_saml2/entityid',
+        get_string('entityid', 'auth_saml2'),
+        get_string('entityid_help', 'auth_saml2'),
+        "$CFG->wwwroot/auth/saml2/sp/metadata.php"
```
shouldn't the default be blank? - eg what happens when a prod site gets set up and then sync to test... you end up with a testing site pointing at prod site metadata.php accidentally....

alexmorrisnz replied:
yep, set to '' now

alexmorrisnz committed:
Add config of more SAML settings

Adds the ability to configure:
- Entity ID
- 'WantAssertionsSigned'
- ACS Bindings
- 'AllowCreate'
- 'AuthnContext'

alexmorrisnz commented:
I've added a new setting 'AuthnContext'

alexmorrisnz commented:
Fixed the behat tests

danmarsden commented on lang/en/auth_saml2.php:
```diff
@@ -108,6 +108,16 @@
+$string['assertionsconsumerservices_help'] = 'List of bindings the SP should support';
+$string['entityid'] = 'Entity ID';
+$string['entityid_help'] = 'Unique name of the entity provider. Leave blank unless required';
+$string['allowcreate'] = 'Allow Create';
```
camel case again :-)

danmarsden commented:
@alexmorrisnz - I'm not sure about the way you're setting the acs.Bindings default values?

I note that the simplesamlphpdocs list 5 options:
https://github.com/simplesamlphp/simplesamlphp/blob/5a705f4dba77b8aa37d3600313b49f6c970b1ef4/modules/saml/docs/sp.md#options

but you've only added 4 and you've set them all to be used by default which would potentially stop the 5th option from workingno idea if that mising one is used now, but if not, I'm guessing it might be used in future?

I'm tempted to suggest we leave the default value for this blank (nothing selected) and then only set the acs.Bindings param when that value is not empty - that way the library will be able to use "all" available bindings rather than just the ones set in the list - this might be important in future if simplesaml upstream library adds new binding options too.

alexmorrisnz commented:
Hi @danmarsden thanks, I wasn't aware of the 5th. I've added it as an option now
I agree with leaving it blank, makes sense. I've made the change

alexmorrisnz commented:
Rebased on the latest MOODLE_39_STABLE branch
